File: go/parsers/abort.go
package parsers

import (
	"google.golang.org/protobuf/proto"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-protobuf/go/gen"
)

type Abort struct {
	gen *gen.Abort
	ex  *PayloadExpander
}

func NewAbortFields(g *gen.Abort, payload []byte) messages.AbortFields {
	return &Abort{
		gen: g,
		ex:  &PayloadExpander{payload: payload, serializer: g.GetPayloadSerializerId()},
	}
}

func (a *Abort) Details() map[string]any {
	return map[string]any{}
}

func (a *Abort) Reason() string {
	return a.gen.Reason
}

func (a *Abort) Args() []any {
	return a.ex.Args()
}

func (a *Abort) KwArgs() map[string]any {
	return a.ex.Kwargs()
}

func AbortToProtobuf(abort *messages.Abort) ([]byte, error) {
	msg := &gen.Abort{
		Reason: abort.Reason(),
	}

	payloadSerializer := selectPayloadSerializer(abort.Details())
	msg.PayloadSerializerId = payloadSerializer
	payload, err := serializers.SerializePayload(payloadSerializer, abort.Args(), abort.KwArgs())
	if err != nil {
		return nil, err
	}

	data, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeAbort, data, payload), nil
}

func ProtobufToAbort(data, payload []byte) (*messages.Abort, error) {
	msg := &gen.Abort{}
	err := proto.Unmarshal(data, msg)
	if err != nil {
		return nil, err
	}

	return messages.NewAbortWithFields(NewAbortFields(msg, payload)), nil
}

File: python/wampprotobuf/parsers/abort.py
from typing import Any

from wampproto.messages import Message
from wampproto.messages.abort import IAbortFields, Abort

from wampprotobuf.gen import abort_pb2


class AbortFields(IAbortFields):
    def __init__(self, msg: abort_pb2.Abort):
        super().__init__()
        self._msg = msg
        self._args = None
        self._kwargs = None

    @property
    def details(self) -> dict[str, Any]:
        return {}

    @property
    def reason(self) -> str:
        return self._msg.reason


def from_protobuf(payload: bytes) -> Message:
    result = abort_pb2.Abort()
    result.ParseFromString(payload)

    return Abort(AbortFields(result))


def to_protobuf(abort: Abort) -> bytes:
    result = abort_pb2.Abort()
    result.reason = abort.reason

    return result.SerializeToString()
